### ISSUE TITLE
ESP 3D 3.0 Status

### ISSUE THREAD

**Issue by luc-github** [luc-github/ESP3D#239]
Ok time to prepare 3.0

What is the plan ? Do what was not done in 2.0/2.1 :wink: 

1 - <s>Nextion Screen support ((ESP3D + Nextion FW)</s> (postponed / cancelled) use touch tft  instead
2 - _M408 support according target FW support_ (WebUI)  (ongoing)
<s>3 - Reprap support</s>
<s>4 - MKS TFT WiFI support (Simulate MKS-WiFi module behavior to use the connector)</s>
<s>5 - ESP3D as Library as part of main board FW (eg Marlin)</s> Code is emmbedded in Marlin ESP3DLib  or GRBL_ESP32 firmware so will be a porting when 3.0 stand alone is done
_6 - UI improvement :
       * Follow disconnection in authenticate page (WebUI + ESP3D) (done :+1: ) 
       * add toshiba flash air support (WebUI)  (done :+1: ) 
7 - Add ethernet support for ESP32 (like olimex)(done :+1: )
8 - Add optional translation file on flash FS for oled ESP3D message like (connected, update, etc...)
9 - Monitor temperatures on Oled in autonomous way (no web browser connected)
10 -Embedded Host support for ESP32 (on hold}
11 - New webUI   (on going)
12 - Fix issue with ESP8266 and Async Webserver and big Serial flow (on hold as 3.0 is planed to be sync only)
13- Put all GCODE in preferences file for easier customization (WebUI)
_14 - Heart beat generation to help UI to known if ESP3D is still connected or not (already implemented in GRBL-ESP32 and Marlin-ESP32 but disabled as buggy)  (WebUI)_ (done :+1: ) 
15 - Fat FS support as option for SPIFFS for ESP32 , LittleFS as option for SPIFFS ESP8266 (done :+1: )
16 - Serial Bluetooth for ESP32 (done :+1: )
17 - Macro autostart (from FS done :+1: ) (from SD : TBD)
18 - Date Time on FS (SPIFFS / FAT / LittleFS) for ESP32 as ESP8266 (done :+1: ) 
Note: ESP8266 SPIFFS time support is not yet ready in core
19 - Add LittleFS support  for ESP32 (TBD)
20 - Buzzer support (done :+1:)
21 - Push Notifications (Pushover / email) (done :+1:)
22- Both way websocket communication  (done :+1:)
23 - Add auto notifications like https://github.com/luc-github/ESP3D/issues/351 (done :+1:)
24-  USB host support using arduino usb host shield (https://github.com/felis/USB_Host_Shield_2.0) (on hold)
25 - Embedded camera support (done :+1:)
26 - SD access with time support: Native (all) / SDFat (all) / SDIO (esp32 only) (done :+1:)
⚠️ Note: this not Printer SD so printer has no access to it, it is for host feature, storage purpose or FW flash purpose.
27 - TFT touch (on hold)
28 - USB Disk support (using CH376s) (not started)
29 - FTP Server (done :+1: )
30 - telegram notification support (done :+1: )
31 - BTT tft support  (on going)

**Comment by Amtho** [comment]
Nice list of work. Sorry i think  i cant help you much to accomplish all. Im more from the electrician site...
But as you my have noticed ;) SD Card support directly on the ESP32 would be my top wish ;)

**Comment by luc-github** [comment]
No worry, testing and feedback are great help, so you help ^_^

**Comment by luc-github** [comment]
PR for sync webserver (esp8266/ESP32) have beed done - first one is in waiting list for 2.5 and second one is merged already

I am spending some time on UI to get consistent UI across browsers and devices - especially mobile devices - seems response behvior of bootstrap 3.0 is little bit messy sometimes - so need to be managed 
- unfortunatly fix depend of each control and position - so need some time - but results are promising so should be ok next week

**Comment by luc-github** [comment]
About Marlin library code is under early development here : https://github.com/luc-github/Marlin/tree/esp32 because hard to separate both code    
Latest Status is here:  https://github.com/luc-github/Marlin/issues/1   
it is alpha state and need code cleaning as I currently I focus on 'feasability' first - once everything is working / code will be cleaned for sure

**Comment by vivian-ng** [comment]
Another feature request (wishlist?): display the interface on a touch screen (not just the Nextion) so that it can be used with generic TFT screens which we usually use with Arduino or STM32 dev kits. This is probably quite a major undertaking with a total design of the menu tree plus touch interface. Maybe start small without touch (use buttons or rotary encoder for inputs). Just an idea that makes the ESP3D effective the "OS" for MKS TFT alternative.

**Comment by luc-github** [comment]
Currently Oled screen is supported 
Having TFT or Nextion make sense only if ESP3D become a Host or you have something else in mind ?

So first step is to add full host feature - which is planned then TFT 

Do you have TFT screen to recommend ? One you already used and happy with 
Nextion is nice for easy customization but I am not really happy with touch responsiveness - so TFT support should happen  before Nextion

**Comment by vivian-ng** [comment]
@luc-github Yes, full host feature is what I was thinking about. The web interface is great already, but without a local interface, it is still somewhat hard to use. So a screen and input (rotary encoder or buttons) would be great. But that only comes after full host feature.

As for screens, I have a few lying around, using ILI9341 (need additional 8 pins), ILI9225 (SPI), ILI9486 (need additional 8 pins), and ST7735S (SPI). I also have a M5Stack, which comes with its own screen and 3 buttons (so it already has all the stuff required to be a host). I am not a very good programmer, but I hope to help by testing and debugging.

**Comment by luc-github** [comment]
@vivian-ng thanks for the feedback - I have ordered a couple of ILI9341 seems well supported by ESP32
I will received them in 2 weeks I guess - I will start to check then

**Comment by luc-github** [comment]
[GRBL-ESP32](https://github.com/bdring/Grbl_Esp32/tree/WebUI)  and [Marlin-ESP32](https://github.com/luc-github/Marlin/tree/eps32_webui) are now using using  [ESP-WEBUI 2.1](https://github.com/luc-github/ESP3D-WEBUI/tree/2.1)
I will create an ESP3D 2.1 branch soon to sync latest bug fixes and features to ESP3D itself

2.1 is mostly ESP32 focused because more and more possibilities / features come with ESP32
the top 2 I am exited to integrate :
 1 - Touch Screen TFT for ESP32 as host (got ILI9341 screen to play with)
2 - Camera (just got  couple of OV2640 modules and M5Stack ESP32 camera modules )

**Comment by luc-github** [comment]
Ok I have re-ordered TFT as the ones I got were without touch :disappointed: 
this time the Touch is confirmed to be XPT2046 spi port ....

**Comment by luc-github** [comment]
I rename topic 3.0 - I am rewriting the complete core because I want to add more cross platform functions and simplify the code by splitting files by functions
So it will be easier to add new features or modify existing ones
because it is big rewrite - I thing it make sense to change the major version number

**Comment by vivian-ng** [comment]
> Hi Luc, the M5 stack camera modules look great, I also found these, and was wondering what you think of them as candidates of ESP3D. The thing that catches my eye is the built in SD card. This opens a few possibilities for host function, or at least file upload to sd card.
https://www.aliexpress.com/item/1pcs-ESP32-CAM-WiFi-WiFi-Module-ESP32-serial-to-WiFi-ESP32-CAM-Development-Board-5V-Bluetooth/32955484091.html
These cameras are not as easy to flash as they do not come with USB port.

Even the new M5Cam with PSRAM is working quite hard to stream the video. Without PSRAM, video streaming will have lag. I am not sure if it can handle the additional load of running ESP3D if video streaming is already such a burden. Instead, I intend to use my ESP32CAM and M5CAM modules to monitor my prints. They can be configured to connect to the home router, and then it is just a matter of inserting "http://192.168.x.x/jpg_stream" as the link for camera streaming in Octoprint or ESP3D.

**Comment by luc-github** [comment]
@vivian-ng ok you confirm my fears

**Comment by luc-github** [comment]
there a recent sample add : https://github.com/espressif/arduino-esp32/tree/master/libraries/ESP32/examples/Camera/CameraWebServer
it may worth to try it

**Comment by just-jason** [comment]
Thanks for the tip, I am still waiting for my board to come from china, could be a while still

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

**Comment by luc-github** [comment]
Quick note, 3.0 have several settings simplifications:
- no more sleep mode configuration : always set to none as connected to printer
- no more PHY configuration : esp8266 looks have now hard coding N mode so no need to bother with useless setting
- AP IP is only static and the only settings that can be configured are IP and channel, masque is set as 255.255.255.0 because only 1 client is supported at once.
- AP is always visible and security will be set to WPA if password is defined or OPEN if no password, so no more settings for SSID visibility neither authentication type.

Feel free to comment if any issue

**Comment by luc-github** [comment]
I am starting working on Ethernet support - I have a board for ESP32 (olimex gateway) but none for ESP8266 .
* Anyone know an existing solution for ESP8266 or one that does not need to solder component ? - I am not good at it :cry:
* Does another board with ethernet exist for ESP32 ? I know there is a POE version but looks expensive.

So currently only the ESP32  Olimex gateway should be supported.

Ethernet  will have DHCP / Static IP and can be act as DHCP server, by default not
So network modes will be :
1 - Nothing
2 - Wifi AP (static IP only)
3- Wifi STA (static or DHCP IP) 
4 - Ethernet Client (static or DHCP IP) 
5 - Ethernet Server (static IP only)
6 - Serial Bluetooth
I think it should cover all needs, to clarify acting as gateway (WiFi / Ethernet) is out of current specs so far.
Feel free to comment

**Comment by luc-github** [comment]
I see several libraries for ILI9341 and Touch XPT2046 SPI 
-> TFT_eSPI (esp8266/esp32)  https://github.com/Bodmer/TFT_eSPI
-> ESP32_TFT_library (esp32) https://github.com/loboris/ESP32_TFT_library

I did not play with any yet

Looks like TFT_eSPI  is multiplatform which would be easier to implement
Anyone used these libraries or any others ones ? Any comment ?

**Comment by vivian-ng** [comment]
@luc-github 
I can see merits of both, so I am quite torn over which to suggest too.
-> TFT_eSPI (esp8266/esp32) [https://github.com/Bodmer/TFT_eSPI](https://github.com/Bodmer/TFT_eSPI)
Bodmer seems to be very active in development of drivers for displays, so it is likely that this library is going to be stable and capable, and can probably rely on Bodmer to help if problems are encountered.

-> ESP32_TFT_library (esp32) [https://github.com/loboris/ESP32_TFT_library](https://github.com/loboris/ESP32_TFT_library)
I thought I recognized the name loboris, then I saw the repos and yup, this is the same person who has a fork of MicroPython working on the ESP32 with PSRAM. So it is again someone who can be relied on to help if any problems are encountered.

Similar to you, I think TFT_eSPI seems to be easier to implement, since it is an Arduino library (ESP32_TFT_library doesn't have any Arduino examples, so it is going to be a bit harder to try an figure it out). It also looks to be easier to configure for different displays through the User_Setup_Select.h header file. This makes the implementation more high-level, and we just need to worry more about how to implement features, and let the library handle the backend low-level driver portion. End users will then just need to change the config file to select their specific display, and can then compile.

**Comment by DerKrawallkeks** [comment]
I just wanted to tell you how great this project is, since I just installed it with an esp8266 on my printer.
Again, is there any chance to increase the speed of upload to SD?

**Comment by luc-github** [comment]
@DerKrawallkeks I have just created a FAQ for this question : https://github.com/luc-github/ESP3D/issues/292, please have a look

**Comment by luc-github** [comment]
I am now rewriting the web events part to use same api as async with sync webserver - this would simplify the ESP3D and ESP3D-WebUI code not using the Websocket for sync instead, it will also be implemented in embedded page for communication control

**Comment by luc-github** [comment]
this year for sure, ASAP I can do , now an exact date - no sorry - I do others projects in parallele and 
3.0 is a big rewrite which also include the WEB-UI, currently I am working on web events, then will implement WebSockets for printer feedback , once done I will work on Embedded Web page to support new communication way, then update webUI to support new communication way also.
Then will add SD support, and also time server for file upload, then will be back to webUI to check if everything is Ok, then will add host support and again will have to add this feature to WebUI

So as you see many to be done before something usable - no mention DHT support and TFT screen support and some others minor features I need to work on

About ESP32 any would fit I think,  I have several ones - and will I try all during test - do not use the `too cheapest chinese ones` - I got some really cheap ESP8266 / ESP32 which were actually really crappy  and not working...
Choose the one you are confortable with , with SD reader is convenient butyou may need the check what SPI pins are used, if you buy without SD you need to buy a SD card reader but this one can have a SD detect feature which is may not the case of embedded ones, but with external SD read it is also easier to use the right pins for connection

**Comment by IInickII** [comment]
Thank you for reply. Can you give me some link on ebay or somewhere for type of board example which you use?
Thanks

**Comment by luc-github** [comment]
https://www.aliexpress.com/wholesale?spm=2114.search0104.8.11.2951ba86EQdEqQ&initiative_id=QRW_20190124063635&SearchText=t8+v1.7+ttgo+esp32&productId=32853967655

They also have ESP32 with 128Mb flash (16MB) which allow to use FAT ( instead of SPIFFS 11MB)

I usually buy from them : https://www.aliexpress.com/store/1983387?spm=2114.search0104.8.7.46a71368F1Pioi

**Comment by luc-github** [comment]
But now it is Chinese New year so I think nothing will be sent before 3 weeks

**Comment by luc-github** [comment]
Testing ILI9143 + ESP32, so far so good ^_^

Just need to solder the SD card reader header, connect all wires, and compile sample accessing JPEG from SD and render
![image](https://user-images.githubusercontent.com/8822552/52001427-ab220e00-24bf-11e9-8362-fd4e877310a1.png)
![image](https://user-images.githubusercontent.com/8822552/52001469-c856dc80-24bf-11e9-9c26-8d986e9797f6.png)
Everything work out of the box - very nice - back screen is little bit too hot (the LED part) almost burned my fingers...
Touch screen reactivity is better than nextion IMHO, but precision is just Ok - need big buttons I think - the keypad sample and calibration sample are working as expected.
The embedded SD reader is working well - just miss an SD detect pin, but can use without it
 
Now need to read  ESP32_TFT API and write some code :smile_cat:

**Comment by luc-github** [comment]
Another change in 3.0 as be a library seems not really a need - be back to single app - and as I got several people not able to install external libraries - everything will be included 
files organization has been also modified to become simple (1 ino file, 1 configuration file, all others files are in subdirectories - including necessary libraries
Current structure as of today (still miss telnet, websocket, oled, tft, host, etc...) :
![image](https://user-images.githubusercontent.com/8822552/52003873-ffc88780-24c5-11e9-8eea-0db630ad0cac.png)
This should avoid end user installation issues (I hope) :crossed_fingers:

**Comment by rikaige** [comment]
Hi Luc,
I'm writing you about Nextion display integration...
In these days I'm starting to develop something new for my RADDS+Due running MalinKimbra firmware, I have a Nextion display on my shelf but don't really like the MK graphics so I'm drawing a new interface and then I'll make them communicate... 
Looking for some WiFi implementation I found your project and your will to implement this display that sounds really cool to me :) because this will be totally cross-firmware&hardware compatible, I have never worked with ESP mcu but I will be happy to help with the graphics and Nextion coding ;)
In the meanwhile I ordered a couple of ESP (8266 & 32) boards to test the WiFi connectivity and look into the code!

Rikaige

**Comment by luc-github** [comment]
Thank you - I won't recommend to use ESP8266 as host + screen support + wifi with all ESP3D feature, not enough power - ESP32 is the board I will test only 
Not acting as host would be a solution for esp8266 but lot of people feel disapointed when they try the SD upload which is slow like hell because of current protocole not really file oriented, acting as host would avoid this.

Sure every  help is welcomed - I am definitly not a good designer - I started to write Nextion UI some time ago :
![nextion1](https://user-images.githubusercontent.com/8822552/52174653-6b954380-2797-11e9-9bc0-d2050e7cf03a.png)
![nextion2](https://user-images.githubusercontent.com/8822552/52174654-6b954380-2797-11e9-99b5-071cf505a393.png)
![nextion3](https://user-images.githubusercontent.com/8822552/52174648-6afcad00-2797-11e9-9036-881a748ef85a.png)
![nextion4](https://user-images.githubusercontent.com/8822552/52174649-6afcad00-2797-11e9-9c5c-15fa65237f67.png)
![nextion6](https://user-images.githubusercontent.com/8822552/52174650-6afcad00-2797-11e9-9528-b90f59b9b111.png)
![nextion7](https://user-images.githubusercontent.com/8822552/52174651-6afcad00-2797-11e9-9310-faae06d3a4fb.png)
![nextion8](https://user-images.githubusercontent.com/8822552/52174652-6afcad00-2797-11e9-9c9c-359380f718f8.png)


 but result is not really nice, so I moved on to other topics until I get an UI I like and so work on its implementation.

I am also starting to work on ili9143 Touch screen implementation, and design will become soon the problem, I am just at reading API and testing code, I  wish to use same UI for Nextion and TFT

What Nextion screen do you have ?

**Comment by rikaige** [comment]
I have a 3,5" 480x320 basic display now, but it was meant to be used in an other project so when I'll have defined the graphics I'll think about the proper screen size, even if it this seems big enough to me...
The cool thing about Nextion, or similar displays, is it is self-sufficient and needs only serial communication to interact with it, thus I was thinking , but not verified yet, about retrieving data using GCode commands as any other host, but leaving all the printing control on the printer hardware, more or less as the ESP is doing I guess... 

Do you think having the ESP wired to the printer hardware on Serial and the Nextion on another ESP Serial (HW or SW) with the ESP that handles the display informations will be too demanding for ESP?

As soon as I have something ready I'll share some screenshots of the idea.

**Comment by luc-github** [comment]
No Nextion is jus at display - it cannot be used as standalone unlike MKS-TFT which is actually a M0 +TFT 
* Nextion use it's own protocol using Serial - so it cannot send pure GCODE, https://www.itead.cc/wiki/Nextion_Instruction_Set, it means ESP must be the MCU handling the code
you can check Nextion library 
* its memory is very low, the minimum to display output - for example file listing cannot be handled directly by nextion, it need to be splited list like 5 names by 5 names
* Nextion SD card is only for flashing screen, no direct access

The Nextion strength is the ease of use to change UI - you can add code the UI to simplify the back end and so the communication with ESP but not so much actually.

Yes the ESP must be between Printer and Nextion - ESP32 have Serial + Serial 1 (serial 2 seems having hardware bug for what I read)
if no host ESP8266 must use Software Serial but I would prefer to start with ESP32 to be sure is working first and see if it is can be optimized for ESP8266 after

**Comment by just-jason** [comment]
Hey Luc , how about some beeps from the printer eg, when commanded to start printing, file upload starts / ends? I am sure you can think of a few more good ideas .

**Comment by luc-github** [comment]
Well could be an idea. But need to be an option as not all Printers/FW support M300 command. Also these commands are time consuming so if use music instead of few beep. Another solution would be to add buzzer on ESP, it won t bother printer with additional command and will work as stand alone what you think ?

**Comment by just-jason** [comment]
I agree , options are always great. I would be happy to have the option for the printer buzzer. Every one can turn in on/ off according to their setup.

**Comment by luc-github** [comment]
Just did a sync with my repo to add support for DHT ,  Reset Pin and Time management https://github.com/luc-github/ESP3D/commit/b0375a3541eca974f88302e5568d71782a92651f
Next modules on a way :
* Notifications pushover and email
* Basic host for Macro and autostart
* oled + TFT 
* rotary encoder + touch

**Comment by dremeier** [comment]
sorry, but where do I find the ESP3D 3.0? I´ve no luck with 2.0.

**Comment by luc-github** [comment]
3.0 is in 3.0 branch but not yet usable, not in pre alpha  as I did not rewrote the webui yet - current webui is not compatible as I have rewritten lot of things. I am first working on pure FW part
I have recently pushed the  notification port for pushover/Line/Email
Next step is the basic host and autostart, then Oled support
if you have issue with 2.0 please open issue

**Comment by dremeier** [comment]
Ok, thank you will fight with 2.0 :-)

**Comment by luc-github** [comment]
3.0 will be nice I think, but I need time to work on it,  sorry

**Comment by aganov** [comment]
@luc-github Do you have any plans for the web ui. My proposal is to use webpack for the build process and then we can use https://github.com/developit/preact (3kb) or https://github.com/choojs/choo (4kb) to organise the UI into components and bring some reactivity. Also we can replace glyph icons with some svg based icons like https://feathericons.com/ for example... What do you think?

**Comment by luc-github** [comment]
Not yet I was planning to reuse the existing javascript I have wrote as I do not want online dependency in case printer is not connected to internet - but I am open to new solutions
I never used these projects do you ?
about glyph currently I use svg extracted from bootstrap, but I am also Ok to change - and there was a request to be able to add new ones so I am fine as soon the size is kept small

To sum up I open to everything but I am not sure I can study all solutions by my own, especially because I do not know them so any proposal is welcome

**Comment by luc-github** [comment]
@aganov I have created https://github.com/luc-github/ESP3D-WEBUI/issues/47 - which explain also size constraint

**Comment by just-jason** [comment]
> I am starting working on Ethernet support - I have a board for ESP32 (olimex gateway) but none for ESP8266 .
> 
> Anyone know an existing solution for ESP8266 or one that does not need to solder component ? - I am not good at it 😢

If you are still thinking about this , then take a look here

https://esp8266hints.wordpress.com/2018/02/13/adding-an-ethernet-port-to-your-esp-revisited/

**Comment by luc-github** [comment]
Yes I know this one - it need to do some soldering ^_^, I addition of not beeing good at it - I am lazy, that is why I was looking for a ready solution 😉

**Comment by luc-github** [comment]
LittleFS is now in esp8266 core git - it should allow time/date unlike current esp8266 SPIFFS, but need to check if current implementation allow it - TBC

**Comment by luc-github** [comment]
about **USB host support using arduino usb host shield** I have ordered a couple of shields to check - they will arrive in few weeks in theory

**Comment by schorhr** [comment]
> about **USB host support using arduino usb host shield** I have ordered a couple of shields to check - they will arrive in few weeks in theory

Hello :-) I am just curious if it works, no rush or anything. Would be amazing. Thank you for your hard work!

**Comment by luc-github** [comment]
@schorhr I have tested a little but seems my hardware is a clone - no wonder it was cheap, and USB connector is not really reliable, so I got a lot of random disconnection , undetected connection, etc...
 I have ordered an uno shield usb host too compare behavior but did not have time yet.
during my test I realized the code need to match the targeted device, which will make task little bit more complex than I expected.
I have also received today some UDisk boards (USB connector with CH376s chip) to read USB drive like SD card, I plan to support it also

**Comment by schorhr** [comment]
Thank you for taking the time to update on the progress!
Too bad you're having troubles with the clones. But I suppose it's a good thing, as many people would use clone boards as well, so it's no surprise later on.
Great! Good luck!

**Comment by ithinkido** [comment]
Hello Luc,

Would it be possible to get the actual flow, feed, ( and fan speed ? ) values? At the moment , if for any reason you reconnect, then the value shown will be 100% even when the machine is running a different value

**Comment by luc-github** [comment]
Which gcode get these values ?

**Comment by luc-github** [comment]
Tft screen have colors support and better resolution than oled so I will integrate littlevgl support. This great library allows very nice UI, for oled I will keep current simple UI
I have already wrote littlevgl drivers to access Flash (Fat and SPIFFs) and SD for images. But in case of flash image storage,  I think 8M  to 16M flash will be necessary, TBC

**Comment by labotecno** [comment]
hello what stuff do you use for your new esp3d 3  ? 
to plain to byy it soon 
i use esp3d 2.1 
what is the last version of your firmware esp3d 2.1 
thank you

**Comment by luc-github** [comment]
any ESP should fit but better have at least 4M flash
depending on needs ou choose ESP32 or ESP8266, some feature like host or even Touch/tft ESP32 is better choice more pins and 2 cores, more feature need more power = esp32, limited/specific feature need less power = esp8266.

**Comment by luc-github** [comment]
Ok after nice POC of https://github.com/eben80/ESP32-CAM-Print-Monitor I took some time to check and was able to make the camera streaming server to coexists with webserver and  ESP3D.
decreasing the task priority and disabling the face recognition helped a lot ^_^
Still need to do some stress tests to verify - but I was able to upload a small file on SPIFFS when video was streaming
There video is streamed using httpd server which require 2 ports available, the commands to modify camera settings are done using [ESP172] command
WebUI is not done yet but camera stream is visible using browser `<ip>:<camera port>/stream`

**Comment by vivian-ng** [comment]
@luc-github Interesting idea for the camera.

I actually have a ESP32-CAM module (actually, a few) which I configured to do a MJPEG stream for the "Camera" tab in ESP3D now (running on Marlin-ESP3D). Of course, the camera module is not connected to my MRR ESPA board except for power, but it would be interesting if I can modify the "webcam" sketch to monitor the serial port for similar [ESP172] codes. Your `Camera::command` method looks like a good place to start.

**Comment by eben80** [comment]
Glad that this device is getting a second look :-)
I'm still using it with my basic functionality, monitoring, manual gcode entry, abort sequence and remote power down. Following you with interest!

**Comment by luc-github** [comment]
@vivian-ng using current 3.0 should be ok out of the box for what you describe , for adding ui command should need to modify the html page of webcam to change the commands to send to webserver for camera- I will try to do it this week as sample.
@eben80 yes thanks you for showing the  possibilities ^_^, I still need to do a compatible WebUI with and without overlay info so it is not currently as convenient as your solution.

**Comment by vivian-ng** [comment]
@luc-github Let me play around with @eben80's sketch and see what I can do with it.

@eben80 I see that your cpp file has a cmd_handler which seems to allow the camera settings to be changed on the fly. If you can expand a bit on how to do that in the README, that would greatly help. Of course, one can always read the code... 😅

**Comment by eben80** [comment]
@vivian-ng that cmd_handler actually comes from the espressif CamWebserver code.
I've remove it from the interface HTML of course by you can still set the values based on the HTML parameters that you can find in [here](https://github.com/v12345vtm/esp32-cam-webserver-arduino-simplified-arduino-html/blob/master/app_httpd.cpp)
In the original code from espressif it is HEX encoded and gzipped.
So far I've found that the most usable frame rate I could get was with SVGA res.
I'm still having issues where the stream stops every once in a while, but the device is still responsive to cmd_handler commands.  Then the only way to recover the stream is rebooting the board through the cmd_handler. Not sure what is happening in that case.
Currently I'm trying to implement the auto-scrolling responses for example when I issue M21 for file listing... but I'm having difficulty reading the [multi-line serial responses](https://forum.arduino.cc/index.php?topic=636487.0) 
I know @luc-github has a very nice implementation to do this but it's not simple enough for my limited skills to re-use :D

**Comment by luc-github** [comment]
sample page https://github.com/luc-github/ESP3D/blob/3.0/esp3d/data/esp32cam.html
with auto detection of stream port and [ESP172] command support
which give :
![image](https://user-images.githubusercontent.com/8822552/65876729-82d71000-e3bc-11e9-8e7e-add6b9ad1a28.png)

the image is full of random line it may be due to  EMI noise on my wires, because even on sample code I have them.

**Comment by eben80** [comment]
@luc-github those look like EMI noise lines yes :) I've seen them before...

**Comment by ETE-Design** [comment]
Would like to try this on my ESP32-Cam module, so I just use ESP3D 2.1 and UI 3.0?
Another thing, is it possible to send files to the SDCard on the module, to speed Up Wireless filetransfer?

**Comment by ETE-Design** [comment]
So FW 3.0 with UI 2.1 to test it?

**Comment by luc-github** [comment]
as wrote there is no UI for 3.0 yet . 2.1 is not compatible
that is why for testing cam I did sample page

**Comment by ETE-Design** [comment]
Mabye this could be ysed for the Cam Support: https://github.com/jameszah/ESP32-CAM-Video-Recorder

**Comment by luc-github** [comment]
SDcard management iimplementation is on going, 
Add FTP server support  should be next addition

Piicture capture and video capture to SD or USB disk will be done later but implemeted for sure

**Comment by luc-github** [comment]
I have updated the status
Recent update : 
SD FS is now supported : native (all) / SDFat (all) / SDIO (only esp32)
The SDFat has specific feature, it allows to Format SD, this would help to avoid unsupported formating
This is for SD card connected to ESP only - printer has no access to this SD card, this will be used by host feature, as storage and way to flash / configure ESP3D. These features are ready but not yet implemented in this repo

**Comment by ETE-Design** [comment]
Any news on the GUI part of this amazing work? Would like to "Test" it soon :-)

**Comment by labotecno** [comment]
i want to test it ! can you tell me when you have the GUI ?

**Comment by luc-github** [comment]
wow so much expectations ^_^
Well I am super late on it - 6 month late already and no defined date yet
Coding the gcode host part give me some challenge and add some delay as I was planning to start after but I may be not wait for host part after all.

In same time I need to build some sytems for Marlin_ESP32 and GRBL_ESP32 and for an ESP3D with esp32  as I am currently coding/supporting them without real system.

I am building them with all spare parts I have,  so it need time to get everything up to work, 

In same time I do support on existing projects and help on another using reprap FW.

Finally, IRL I help my aunty to setup a new house by building furnitures/doing small works.
 
So let say I have no excuse but being super busy :D

Please be patient, I cannot give a clear date yet sorry, but will work on it soon (in theory)

**Comment by ETE-Design** [comment]
@luc-github No expections, just looking forward to try the Esp32-Cam - That's my reason for asking ;-) Can hear you have a lot to do since you have to help all the others also...

**Comment by labotecno** [comment]
@luc-github  thank you to much ! can i use the normal version with esp32 ? je peux utiliser le GUI 2.1 avec esp32 merci

**Comment by luc-github** [comment]
GUI 2.1 + ESP3D 2.1 + ESP32 = Ok
No compatible GUI for 3.0 for the moment but sample pages

**Comment by ETE-Design** [comment]
@luc-github So I can flash 3.0 on the ESP32-Cam and use a sample page to see output? Is there a how to for that?

**Comment by luc-github** [comment]
flash as usual - configure network using [ESPXXX] commands:
```
[ESP100]<AP to join>
[ESP101]<AP password>
[ESP102]DHCP
[ESP110]WIFI-STA
```
the uploader page is available so you can upload the sample page for camera

**Comment by luc-github** [comment]
if BTT FW works great on MKS TFT  I may not porting from 2.1:
`4 - MKS TFT WiFI support (Simulate MKS-WiFi module behavior to use the connector)`
Feel free to share if any concern

**Comment by luc-github** [comment]
I will hold the not yet implemented/partially implemented features on ESP3D 3.0, so these feature will be in next version (V3.1?)
Instead I will work on WebUI 3.0 from now to have at least a full package in 3.0 to be tested.

**Comment by ETE-Design** [comment]
@luc-github So BBT FW Work on MKS Screen? Have to try that then... Perfect you hold implementation for now and Work on the WebUI, is you need help testing just reach out :-)

**Comment by luc-github** [comment]
@ETE-Design have look here https://github.com/darkspr1te/MKSTFT_Marlin_Touch
I did not tested yet

**Comment by luc-github** [comment]
I am more than late on ESP3D webUI so now everything on 2.1 is put on hold so I can focus ESP3D-webUI 3.0

March should 100% be dedicated to new webUI for ESP3D 3.0 - so please do not find any critical bug until new webUI is done 🙏 😹 
thank you for your patience ^_^

**Comment by vivian-ng** [comment]
@luc-github If it is not too much to ask, can there be a toggle switch or checkbox in the main window for 3.0 that allows to toggle camera on or off? Having the printer controls and camera view in one page will really help for remote operations. Just a suggestion. Of course, once everything else on the list at the top is done. :sweat_smile:

**Comment by luc-github** [comment]
@vivian-ng you are refering which camera ?
ip camera ?
or 
esp32 camera embeded with esp3d ?

**Comment by vivian-ng** [comment]
@luc-github Nothing fancy, just the current IP camera support that is already available in the separate tab. I just thought it would be nice to be able to view that same camera in the main tab where all the controls are. That way, you can actually see the printer moving when giving commands for the X, Y, and Z movements.

**Comment by luc-github** [comment]
@vivian-ng  do you know that you can detach the camera window in current webUI ? - which allows of having control and camera windows side by side.

![image](https://user-images.githubusercontent.com/8822552/75235587-c17ec400-57bc-11ea-8937-39e6f6a3b98a.png)


For 3.0 I was planned a "camera mode" in addition to normal mode with camera as main window and some control in overlay (jog / and temperature monitor) like for ESP32-camera, but the IP camera detach mode will also be present like today

**Comment by ETE-Design** [comment]
@luc-github Do you have any ETA on when we can begin testing the v3.0?

**Comment by luc-github** [comment]
no, I have put in hold all others  projects, March will be only for 3.0 web UI, so end of March, I hope to have better visibility. 
the Webui 3.0 is using new concept/langage for me (Preact) so define an ETA is impossible for me, it is a new journey.
The target is ASAP, what ever it take

**Comment by labotecno** [comment]
hello when we can try the last version ? can be mix with version 2 ? webgui

**Comment by luc-github** [comment]
@labotecno :
https://github.com/luc-github/ESP3D/issues/239#issuecomment-558253300
https://github.com/luc-github/ESP3D/issues/239#issuecomment-590810325

3.0 without webUI is functional, configuration can be done using [ESPXXX] commands

**Comment by bkzspam** [comment]
Hi, can you add temperature graphic to webUI? It can be great...

**Comment by luc-github** [comment]
there is already in 2.1 or do you request somesting else for graphical rendering ?

![image](https://user-images.githubusercontent.com/8822552/76163291-26cf9f00-6145-11ea-9166-8cfd5060b78a.png)
need auto check on 

![image](https://user-images.githubusercontent.com/8822552/76163352-9ba2d900-6145-11ea-94ec-34b36be94011.png)

**Comment by bkzspam** [comment]
> there is already in 2.1 or do you request somesting else for graphical rendering ?
> 
> ![image](https://user-images.githubusercontent.com/8822552/76163291-26cf9f00-6145-11ea-9166-8cfd5060b78a.png)
> need auto check on
> 
> ![image](https://user-images.githubusercontent.com/8822552/76163352-9ba2d900-6145-11ea-94ec-34b36be94011.png)

Oh, thank you. I didn't notice that.

**Comment by labotecno** [comment]
hello again can i use esp3d with esp32-cam?  to get the cam working at same time thank you

**Comment by luc-github** [comment]
@labotecno  https://github.com/luc-github/ESP3D/issues/239#issuecomment-536529510

**Comment by labotecno** [comment]
its possible to add the code from this ? https://github.com/eben80/ESP32-CAM-Print-Monitor , only the version 3 have this option ?

**Comment by luc-github** [comment]
yes only in 3.0

**Comment by ETE-Design** [comment]
@luc-github  So nice to see that you Work hard on getting the 3.0 Released... Will it Work out of the box with ESP32-Cam, also Will it upload direct to the ESP32-CAM's SDCard to allow faster transfer rate, and then send it to the printer / printer SDCard?
Also do you need some a tester for it? Have the ESP32-CAM and an Board with a module possible to add it direct... Trough Wire ;-)

**Comment by luc-github** [comment]
>Will it Work out of the box with ESP32-Cam,

yes it is the case now with test web page

> Will it upload direct to the ESP32-CAM's SDCard 

 yes it is the case now with current ESP3D 3.0

>then send it to the printer / printer SDCard

what do you mean ? How to transfer from ESP32_cam SD to Printer SD ?

**Comment by ETE-Design** [comment]
@luc-github Wasn't sure on it would print direct from the ESP32-CAM'S SD-Card...
Will there be a "Preview 3.0" Have a projekt where I would like to test it on...

**Comment by luc-github** [comment]
Host feature is not close to be done as I focus to ESP3D-WEBUI 3.0, so I do not see how to transfert from ESP32_CAM to printer 
Also if host may works on ESP32 - it will like will not really good durring streaming video IMHO but TBD

Current ESP3.0 FW miss the Serial File transfert but others implemented/missing features are in first post  https://github.com/luc-github/ESP3D/issues/239#issue-338305513
What is not on hold  should  be part on 3.0, others features will be postponed to 3.1  unless critical or super easy to add

3.0 may have bugs of course that is why I focus on webUI now to have a `complete solution to be tested properly`

**Comment by ETE-Design** [comment]
So just to be sure, the SD-Card on the ESP32-Cam is just used for Video / Pictures now, and when transfering a G-Code file to the printer using the ESP32-Cam module, it just transfer the file direct to the printers SD-Card?
Question, the connection diagram for the ESP32-CAM show you have to use 5V and the logic level converter for 3.3v, isen't it just possible to use the 3.3v pin on the ESP32-CAM?
Of cause 3.0, it is hard to try on all setups, that's why I asked if there is a test version out for your new WebUI 3.0

**Comment by ETE-Design** [comment]
So just to be sure, the SD-Card on the ESP32-Cam is just used for Video / Pictures now, and when transfering a G-Code file to the printer using the ESP32-Cam module, it just transfer the file direct to the printers SD-Card?
Question, the connection diagram for the ESP32-CAM show you have to use 5V and the logic level converter for 3.3v, isen't it just possible to use the 3.3v pin on the ESP32-CAM?
Of cause 3.0, it is hard to try on all setups, that's why I asked if there is a test version out for your new WebUI 3.0

**Comment by ETE-Design** [comment]
So just to be sure, the SD-Card on the ESP32-Cam is just used for Video / Pictures now, and when transfering a G-Code file to the printer using the ESP32-Cam module, it just transfer the file direct to the printers SD-Card?
Question, the connection diagram for the ESP32-CAM show you have to use 5V and the logic level converter for 3.3v, isen't it just possible to use the 3.3v pin on the ESP32-CAM?
Of cause 3.0, it is hard to try on all setups, that's why I asked if there is a test version out for your new WebUI 3.0

**Comment by ETE-Design** [comment]
So just to be sure, the SD-Card on the ESP32-Cam is just used for Video / Pictures now, and when transfering a G-Code file to the printer using the ESP32-Cam module, it just transfer the file direct to the printers SD-Card?
Question, the connection diagram for the ESP32-CAM show you have to use 5V and the logic level converter for 3.3v, isen't it just possible to use the 3.3v pin on the ESP32-CAM?
Of cause 3.0, it is hard to try on all setups, that's why I asked if there is a test version out for your new WebUI 3.0

**Comment by luc-github** [comment]
no, sd card is available but actually not used, no picture no video saving feature implemented yet
I am not sure what you mean about SD card file transfert to printer
fast upload can be done on esp32 sd because sd is connected to esp32
for printer sd  only serial upload is possible unless there is a way to share the sd
so far 5v is necessary because camera need power, remove camera and it will be ok
you can follow webui progress here https://github.com/luc-github/ESP3D-WEBUI/issues/94

**Comment by ETE-Design** [comment]
I mean, if SD-Card not used on the ESP32-Cam for now, then it isen't possible to do fast transfer? I talk about sending the files to the printer I want to print.

**Comment by luc-github** [comment]
the question is how to prevent   concurrent access to SD
there is no sharing mecanism

**Comment by rhaberer** [comment]
So if I get the host-feature right, you are planning to send the parsed gcode-commands from the gcode-file (stored on esp32's SD-card) directly to the printer while printing?

Have you planned some kind of API for the ESP3D project, so one can access the system without the WEBUI over IP-address and port? Like e.g. uploading, delete a file with that API and start printing.

Impressive work by the way!!!

**Comment by luc-github** [comment]
@rhaberer embedded host won't be in 3.0 but after -  I focus on existing feature to build the webUI, because not so simple to get something fullproof

 the latest status is here https://github.com/luc-github/ESP3D/issues/243 and still under definition

all commands but upload are here : https://github.com/luc-github/ESP3D/blob/3.0/docs/Commands.txt
For upload FTP and HTTP are the only currently supported - webUI is not necessary for them, ftp client or any wget/curl client should work - but not fully tested yet

**Comment by ETE-Design** [comment]
@luc-github Looking at the connection diagram for the ESP32-Cam... It show some resistors, but when you use an logic level converter it refree to the diagram for the D1-mini wiki page and that also use the logic level board for the RX-TX... So what is the correct way to connect it ???

**Comment by luc-github** [comment]
sorry I do niot understand your question
as mentionned : `note: this is not the only way to connect the board.`
* use resistor 
or 
* use logic level converter like for wemos wiki page

**Comment by ETE-Design** [comment]
@luc-github Aha, did not see that I should use one of them only :-) Will an 1Kohm and an 2.2Kohm be close enough, seems hard to find and 2Kohm here in Denmark

**Comment by luc-github** [comment]
the key point is need 3.3v max on GPIO as ESP32 is not 5v tolerant 

so you can use any solution which fit your needs

**Comment by ETE-Design** [comment]
@luc-github Here they say that the ESP32-Cam already have an 5V to 3.3v regulator... If that is true do I then need the 1KOhm and 2KOhm on the RX -> TX?
https://technoreview85.com/how-to-program-esp-32-cam-using-arduino-uno-board/
If that's not the case, is it then possible to just use the voltage reculator Chris uses in the Install Instructions Video you have on your page?
Last, do I need any special plugin for the Sample Esp32Cam to work, cause I can see the menu, but when I press Start Stream I just get an red cross where the stream should have been...

**Comment by luc-github** [comment]
the 5v to 3.3v regulator is to power supply the esp32, I doubt the GPIO have it - did check the GPIO ?
Sorry do not ask me about electronic I am not good at it - that is why I let people share their experience on wiki and success how to connect esp boards

it is same page as esp32cam sample code so check if sample code work with your cam

**Comment by ETE-Design** [comment]
@luc-github Is there an easy way to erase EEPROM om a ESP32-Cam, I'm on Windows... Cause each time I flash it accept it, but do not apply the changes I have made :-(
BTW, until now your new GUI looks great, looking foward to the Camera support, hope you soon implement it in the new GUI, maby just a sample page who can be accessed from the GUI....

**Comment by luc-github** [comment]
you mean ESP3D settings save in EEPROM? or ESP 32 Flash

**Comment by ETE-Design** [comment]
@luc-github Both I think... It is when I change wireless settings nothing happend... Also how to upload new index.html.gz when you have the sample ESP32 camera? There is no Button for that...

**Comment by luc-github** [comment]
changing ESP3D settings in code won't have any effect if you already flashed ESP3D before, you need to reset with [ESP444]RESET to apply them from code to new EEPROM

to access uploader it is like in 2.1
http://your-ip?forcefallback=yes

espcam sample page is only for camera test nothing else so no upload inside, you need the embedded uploader

**Comment by ETE-Design** [comment]
@luc-github Now that I have the sample camera page Up, how to Access the embedded loader?
Also why I asked for the Sample Page "integrated" to your new GUI for easy use ;-)

**Comment by luc-github** [comment]
http://your-ip?forcefallback=yes

**Comment by ETE-Design** [comment]
@luc-github Finaly got it on my network... Also had to do a [ESP710]FORMAT, so now it go to the upload site each time... Problem now is it freeze each time I try to upload a new index.html.gz, it only show "Uploading..."

**Comment by luc-github** [comment]
@ETE-Design it may be a power supply issue - check what is the serial output

**Comment by ETE-Design** [comment]
@luc-github  Have it connected to Arduino Uno 5V, nothing else connected.....

**Comment by luc-github** [comment]
@ETE-Design so?

**Comment by ETE-Design** [comment]
@luc-github Still the same :-( Just show "Uploading..."

**Comment by luc-github** [comment]
same answer sorry
@ETE-Design it may be a power supply issue - check what is the serial output

**Comment by ETE-Design** [comment]
@luc-github 4.84V

**Comment by luc-github** [comment]
i was talking about power stability , esp is sensitive and can crash easily
the serial output would give you crash stack

**Comment by ETE-Design** [comment]
@luc-github  Ohh... Is there an easy way to check that?

**Comment by ETE-Design** [comment]
@luc-github Got it uploaded... So can make tests for you now... ESP32-Cam, ESP3D V3.0, WEBGUI V3.0

**Comment by GerogeFu** [comment]
@luc-github Hi, Luc , would you like this project to have a official supporter? Supply you all hardware parts and other things. You can send me email to talk about this : fzl@fysetc.com.

**Comment by luc-github** [comment]
Thank you @GerogeFu I have emailed you 😸

**Comment by ModMike** [comment]
@GerogeFu It is nice to see Fysetc offer this. I love your Cheetah boards, especially the new 1.1b version! I have 5 so far and give to friends as gifts and prominent youtubers.

**Comment by ETE-Design** [comment]
@GerogeFu @ModMike Mabye it is because Fysetc has just Released a WIFi SDCard they like to Work with it ;-) But nice to see @GerogeFu in here, normal he know how to deal with firmware / hardware :-D

**Comment by ModMike** [comment]
@GerogeFu Sent you a support email.

**Comment by ModMike** [comment]
On another topic, can someone explain to me the big difference between WebUI 2.1 and 3.0? It looks as if 3.0 is a complete controller solution designed to interface with Marlin. Is that correct? What are the possibilities? I saw some talk of using the ESPCAM32 but not sure what that means.

**Comment by luc-github** [comment]
@ModMike 
Let me summarize
Web UI 2.1 and WebUI 3.0 are based on different development way
2.1:
* use pure javascript and several piece of code I stripped and some i wrote from scratch to get the minimum size at the end a
* monolitic GRBL + 3DPrinter + multilanguage  or 2 languages

3.0:  
* is based on PreactJS which work at DOM level - 
 * the GRBL code and 3D printer code are splited and language pack but english are external and additive according needs - this allowing to keep small size at the end when no code stripping is done

The Camera support is another thing it is linked also to ESP3D FW for ESP32-CAM
Original feature of having an external ip camera is still there but in addition it will support esp32-cam with ESP3D, for this the UI can be different using a camera view as main interface with notification informations - streaming video using a lot of power from ESP board and upload won't be possible when streaming for example.  

The UI is not designed for Marlin only but also Smoothieeware and Repetier, it is also designed for GRBL/GRBL_ESP32 and for ESP3DLib 3.0 that will be for Marlin_ESP32 boards, no ESP3DLib 2.1 is planned, jump will be from 1.0 (current) to 3.0, and also Panucatt Devices ESP3D based (WiFi backpacks , Azteeg WiFi, WiFi ViKi LCD) 

Also there are lot of feature in WebUI 3.0 that are not present in 2.1 :
*  progression of printing, 
 * export of settings,
*  mobile support is far better,
*  UI is simplified to give fast/easy access to features,
*  more sensor support,
*  easier Macro creation, others 
* ESP3D devices discovery
* better wizard with profile for identified configurations
* etc...
 
ESP3D 3.0 FW has also lot of new features as you can see in first post:
*  better code organization to make easier addition of new components, 
* FTP feature (which allow user to mount ESP as virtual drive),
*  Global FS for connected devices 
 * new FS support (littleFS, SDFat, SDIO), 
 * ESP32-Cam support, 
 * new sensors support, 
* ethernet support, 
* buzzer support, 
 * new notifications (e.g: telegram), which allow some autonotifications (e.g message sent with IP when ESP 3D is started)
 * Serial BlueTooth support for ESP32, 
 * better timestamp for all FS, 
* Autostart Macro (your request) 
 * even LUA script language support for Macro commands to fit users scenari that are not hard coded in ESP3D

And more to come ^_^ 

I hope it answer your question 😸

**Comment by luc-github** [comment]
I think now ESP3D 3.0 will move to alpha stage as WebUI is now usable even not completed for 3D printer : https://github.com/luc-github/ESP3D-WEBUI/issues/94

**Comment by ithinkido** [comment]
> I am starting working on Ethernet support - I have a board for ESP32 (olimex gateway) but none for ESP8266 .
> 
> * Anyone know an existing solution for ESP8266 or one that does not need to solder component ? - I am not good at it 😢
> * Does another board with ethernet exist for ESP32 ? I know there is a POE version but looks expensive.
> 
> So currently only the ESP32 Olimex gateway should be supported.
> 

> 
> EDIT: Removing Ethernet Server because DHCP server is currently only for WIFI AP not ethernet 😭

If you are still looking at ethernet equipped boards 
https://www.aliexpress.com/item/4001122992446.html?

**Comment by luc-github** [comment]
@ithinkido thank you - I will have a look

**Comment by florin03** [comment]
Hello,

Any update on:

24- USB host support using arduino usb host shield (https://github.com/felis/USB_Host_Shield_2.0) (on hold)

I am really interested in this solution, usb is a lot faster than serial (some times on serial i heave problems with printing large files).

Thx

**Comment by luc-github** [comment]
No - I am currently busy in ESP3D 3.0 test with current features and on WEBUI for 3.0 - if any update it will raise here

**Comment by Bastlwastl84** [comment]
i tried your newest ESP3 alpha Firmware and checked at Config if AI Thinker is enabled. Camera failed. Checked the Pins with the Arduino Example if the same. With Arduino exampe the Cam works, With the FW_VERSION "3.0.0.a59" not. Erased the Chip before flashing with VSCode and also flashed with Arduino. ESP32-S Ai-Thinker
rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0018,len:4
load:0x3fff001c,len:1216
ho 0 tail 12 room 4
load:0x40078000,len:9720
ho 0 tail 12 room 4
load:0x40080400,len:6352
entry 0x400806b8
;Starting Network
M117 Starting WiFi
M117 192.168.0.1
M117 AP SSID: 'ESP3D' is started and protected by password
M117 Captive Portal started
M117 HTTP server started port 80
M117 Telnet server started port 23
M117 Ftp server started ports: 21,20,55600
M117 Websocket server started port 8282
**M117 Failed start camera streaming server**
M117 192.168.0.1
M117 Mouning SD
M21

**Comment by luc-github** [comment]
@Bastlwastl84 I cannot duplicate your issue - so I suggest you to open a real issue ticket with all informations needed for support- this thread is about status  not issue
Currently camera does not like hot start - only cold boot 
FYI my boot and status
```
;Starting Network
M117 Starting WiFi
M117 192.168.0.1
M117 AP SSID: 'ESP3D' is started and protected by password
M117 Captive Portal started
M117 HTTP server started port 80
M117 Telnet server started port 23
M117 Ftp server started ports: 21,20,55600
M117 Websocket server started port 8282
M117 Camera server started port 9600
M117 192.168.0.1
```
[ESP420]plain
````
chip id: 50221
CPU Freq: 240 Mhz
CPU Temp: 44.4 C
free mem: 106.35 KB
SDK: v3.3.1-61-g367c3c09c
flash size: 4.00 MB
size for update: 1.87 MB
FS type: SPIFFS
FS usage: 0 B/169.38 KB
baud: 115200
sleep mode: none
wifi: ON
hostname: esp3d
HTTP port: 80
Telnet port: 23
Ftp ports: 21, 20, 55600
Websocket port: 8282
camera name: ESP32-CAM(4)
camera ports: 9600 - 9601
ap: ON
mac: 24:0A:C4:BB:2D:C5
SSID: ESP3D
visible: yes
authentication: WPA2
DHCP Server: ON
ip: 192.168.0.1
gw: 192.168.0.1
msk: 255.255.255.0
clients: 0
sta: OFF
mac: 24:0A:C4:BB:2D:C4
i-time: OFF
serial: ON
notification: OFF
sd: direct(SDIO)
Target Fw: unknown
FW ver: 3.0.0.a60
FW arch: ESP32
````

**Comment by maf-o** [comment]
> The UI is not designed for Marlin but Smoothieeware and Repetier, it is also designed for GRBL/GRBL_ESP32 and for ESP3DLib 3.0 that will be for Marlin_ESP32 boards, no ESP3DLib 2.1 is planned, jump will be from 1.0 (current) to 3.0, and also Panucatt Devices ESP3D based (WiFi backpacks , Azteeg WiFi, WiFi ViKi LCD)

If i plan to use ESP CAM Modul with 3.0 for Marlin on printer it wont't work?

**Comment by luc-github** [comment]
the meaning is not only for Marlin but also .... - typo mistake
any esp32 with Marlin should work but it is still in alpha - and currently having hard time handling embedded camera because is use a lot of esp power

**Comment by maf-o** [comment]
Ok, so i would give it a try.
Especially if there are only problems while streaming.
If its possilbe to send every 5min a telgram photo and i can sometimes look for livestream everyting would be ok :)

**Comment by luc-github** [comment]
for camera support I am on it you can give try to this : https://github.com/luc-github/ESP3D/tree/camera-test
but I have still some issue - send photo is not done yet / neither time lapse

**Comment by luc-github** [comment]
but webui loose connection time to time when streaming  is on when doing query

**Comment by luc-github** [comment]
/snap function has just been added to take picture in jpeg format 
![image](https://user-images.githubusercontent.com/8822552/96460590-a832f800-1223-11eb-8e21-b2ed639181f9.png)

**Comment by luc-github** [comment]
Streaming from camera is consuming too much bandwidth and time to time (often actually) webui queries are rejected, splitting activities in different tasks do not help - the streaming task keep esp32 busy and multitasking is not doing as expected 
As streaming is actually mjpeg - I have removed the streaming code and webUI will do the periodic queries, so of course it will be less smooth because esp32 will now recover some mcu time to process others  commands but I think it is acceptable to give priority to esp3d tasks instead of only streaming only and no flow control, the webUI should be able to control the flow as it send each commands one by one, the esp32cam.html has been updated accordingly

**Comment by luc-github** [comment]
WebUI and FW code now work with embedded camera and flow control

**Comment by ETE-Design** [comment]
@luc-github Seems great, and can see you already working on WebUI 3.1 with @alxblog... Now that you have a PC + Mobile Version... Mabye you could look at the TTGO Watch 2020... Build on ESP32... Would be so nice if you could build an application for it to control / view settings from ESP3D...
BTW have one on it's way, was hoping to make it work with your ESP3D... And thanks for the great work you do :-)

**Comment by luc-github** [comment]
The esp3d features for 3.0 is here https://github.com/luc-github/ESP3D/blob/3.0/ESP3D-features.xls as you can see we are far from done 
the 3.0 webUI 3d printer is still not done using current code (even starting now rewrite is for 3.1 as it is not a small task) 
there is also https://github.com/luc-github/ESP3D/issues/523 for people complaining about upload speed but I do not have any hardware so I do not know when I could work on it 

then it will be the esp32 tft full support (ESP3D v3.2 ?) 
may be TTGO Watch  could be part of it but to be honest I do not see how to link the 3D printer / cnc to this watch  and I even do not  see the point - I do not understand the user case sorry

**Comment by ETE-Design** [comment]
@luc-github About TTGo Watch, I was just thinking about a piece of software for it to get the info on the watch from WebUI, don't want the watch to be the brain... So just a UI for watch screen's...

**Comment by luc-github** [comment]
@ETE-Design if need to use esp3d info, the best would be to use tcp connection instead of use webUI as starting point 
TFT and webUI are totally different rendering 
I understand the fun of `it could be done` - but I have no bandwith  for such thing,  user case of TTGO watch rendering vs mobile phone rendering is very very very low if even happen
but I guess it could be naturaly done when real TFT support will be done - it will just use tcp connection instead of serial connection to get infos / send commands

**Comment by MillerRen** [comment]
@luc-github Direct SD is supported now? Which one is? TFT SD or TFT USB or Printer SD? I'd like it to drive grbl offline. This is a great project.

**Comment by luc-github** [comment]
SD directly connected to ESP board so none of the ones you mentioned

**Comment by luc-github** [comment]
if you use GRBL - better use this FW https://github.com/bdring/Grbl_Esp32 , the board using it has a sd card reader and so allows direct access

**Comment by luc-github** [comment]
About file upload, to clarify and avoid recurrent questions, here is what the 3.0 should support in beta,  this is an under going and planned work, implementation should follow list order.
 
* UPLOAD_DIRECT_SD (implemented)
the SD is directly connected to ESP board but printer has no access to it
e.g: ESPCAM32 board
       
* UPLOAD_SHARED_SD  (implemented)
the SD is shared between ESP and printer using hardware switch
e.g: Panucatt wifi back pack and other related devices

* UPLOAD_SERIAL_SD   (implemented in 2.1 but not available yet in 3.0)
the access to SD is done using M28/M29 protocol
e.g: Default 3D printer protocol

* UPLOAD_FAST_SERIAL_SD (not started yet,  waiting for hardware)
the access to SD is done using an additional  pin to let printer FW / TFT FW know this is an upload 
e.g: MKS TFT / MKS Robin 
Note: it is currently limited to MKS TFT and boards using Marlin as no one else implemented this protocol yet

* UPLOAD_FAST_SERIAL_USB (not started yet, waiting for hardware)
the access to USB TFT port is done using an additional  pin to let TFT FW know this is an upload 
e.g: MKS TFT 
Note: it is currently limited to MKS TFT as no one else implemented this protocol yet

* UPLOAD_DIRECT_USB   (not started yet, and may be delayed to 3.1, waiting for hardware setup)
the access to direct USB port using CH376s chip
Note: CH376S support is not yet implemented but definitly planned
